yopt/src/main/scala/pravda/yopt/CmdDecoder.scala

```diff
@@ -53,22 +57,94 @@ object CmdDecoder {
+                _ => Left(LongDecodeError(hex).toString),
+                number => Right( (number, line.tail) )
+              )
+            case moreLikelyDec => Try(moreLikelyDec.toLong).fold(
+              _ => Left(LongDecodeError(moreLikelyDec).toString),
+              number => Right( (number, line.tail) )
+            )
           }
         }
         .getOrElse(noValueError)
 
     override val optInfo: Option[String] = Some("<long>")
   }
 
+  implicit val booleanReader: CmdDecoder[Boolean] = new CmdDecoder[Boolean] {
+    override def decode(line: Line): Either[String, (Boolean, Line)] =
+      line.headOption
+        .map { item =>
+          item.toLowerCase match {
+            case "true"  => Right( (true, line.tail) )
+            case "false" => Right( (false, line.tail) )
+            case "yes"   => Right( (true, line.tail) )
+            case "no"    => Right( (false, line.tail) )
+            case "y"     => Right( (true, line.tail) )
+            case "n"     => Right( (false, line.tail) )
+            case _       => Left(BooleanDecodeError(item).toString)
+          }
+        }.getOrElse(noValueError)
+    
+    override def optInfo: Option[String] = Some("<boolean>")
+  }
+
+  implicit val bigDecimalReader: CmdDecoder[BigDecimal] = new CmdDecoder[BigDecimal] {
+    override def decode(line: Line): Either[String, (BigDecimal, Line)] =
+      line.headOption
+        .map(item => 
+          Try(BigDecimal(item)).fold(
+            _ => Left(BigDecimalDecodeError(item).toString),
+            v => Right( (v, line.tail) )
+          )
+        )
+        .getOrElse(noValueError)
+    override def optInfo = Some("<bigdecimal>")
+  }
+
+  implicit val doubleReader: CmdDecoder[Double] = new CmdDecoder[Double] {
+    override def decode(line: Line): Either[String, (Double, Line)] =
+      line.headOption
+        .map(item =>
+          Try(item.toDouble).fold(
+            _ => Left(DoubleDecodeError(item).toString),
+            v => Right( (v, line.tail) )
+          )
+        )
+        .getOrElse(noValueError)
+    override def optInfo = Some("<double>")
+  }
+
+  implicit val uriReader: CmdDecoder[URI] = new CmdDecoder[URI] {
+    override def decode(line: Line): Either[String, (URI, Line)] =
+      line.headOption
+        .map(item =>
+          Try(new URI(item)).fold(
+            _ => Left(UriDecodeError(item).toString),
+            v => Right( (v, line.tail) )
+          )
+        )
+        .getOrElse(noValueError)
+    override def optInfo = Some("<uri>")
+  }
+
+  implicit val durationReader: CmdDecoder[Duration] = new CmdDecoder[Duration] {
+    override def decode(line: Line): Either[String, (Duration, Line)] =
+      line.headOption
+        .map(item =>
+          Try(Duration(item)).fold(
+            _ => Left(DurationDecodeError(item).toString),
+            v => Right( (v, line.tail) )
+          )
+        )
+        .getOrElse(noValueError)
+    override def optInfo = Some("<duration>")
+  }
+
```

Comment:
I would suggest to use type tags, like `Int @ Hex` or `Boolean @ YesOrNo` and implement additional parser abilities only in these implicit instances. 

The next step is to add combinators: `opt[Int |+ Int @ Hex |+ Int @ Binary]` will try to parse `Int` as decimal, hex or binary number _(this point is not necessary for the PR)_.